### PR TITLE
[WIP] Context and Multiprocessing

### PR DESCRIPTION
**Pull request by abdelhakeem**
This carries on the work done in #9943.

I've tidied the commit log ~although I do not guarantee that all commits are in a consistent state, but the current branch state is stable~ (cleaned!).

Next steps for now:

- [x] More coverage
- [x] `&:` command documentation
- [x] More examples in documentation
- [x] Context additions
- [x] Context documentation

### CODE REVIEW

File: src/nvim/context.h [hunk 1]
```diff
@@ -24,21 +25,54 @@ typedef kvec_t(Context) ContextVec;
...
+  kCtxTVars   =   (1 <<  7),  ///< Tab variables
+  kCtxLVars   =   (1 <<  8),  ///< Function-local variables
+  kCtxSFuncs  =   (1 <<  9),  ///< Script functions
+  kCtxFuncs   =   (1 << 10),  ///< All functions
 } ContextTypeFlags;
 
+extern int kCtxVars;
 extern int kCtxAll;
 
+#define CONTEXT_TYPE_FROM_STR(types, str) \
```
Comment (justinmk) [inline src/nvim/context.h]: seems like this is used in only 1 place. better to avoid macros where possible.

Comment (abdelhakeem) [reply src/nvim/context.h]: Two places, `eval.c` and `vim.c`, in general anywhere I want to make a flags variable from some context type strings

**Comment by abdelhakeem** [comment]
I'll split this into two PR's, one for Context and one for Multiprocessing, so that the Context one can get merged soon.

**Comment by justinmk** [comment]
Maybe useful: https://github.com/vim/vim/pull/5211 refactors `ex_vimgrep` so that the args parsing is separated.

**Comment by yashpalgoyal1304** [comment]
hi! what is the status of this?

**Comment by abdelhakeem** [comment]
> hi! what is the status of this?

Hi, @yashpalgoyal1304! It's been stale for quite a long time, also not sure if there is still interest from the community...

This PR was supposed to be split into two PR's, one for the context feature and another for the multiprocessing feature. The context feature PR, which includes some improvements and some additions to the context dictionary, is at [#11029 ](https://github.com/neovim/neovim/pull/11029). The multiprocessing feature PR is yet to be opened.

**Pull request closed without merging**